nexus/src/app/background/blueprint_execution.rs

```diff
@@ -0,0 +1,407 @@
+                            "Successfully deployed zones for sled agent";
+                            "sled_id" => %sled_id,
+                            "generation" => config.generation.to_string()
+                        );
+                        None
+                    }
+                }
+            })
+            .collect()
+            .await;
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+}
+
+impl BackgroundTask for BlueprintExecutor {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            // Get the latest blueprint, cloning to prevent holding a read lock
+            // on the watch.
+            let blueprint = self.rx_blueprint.borrow_and_update().clone();
+
+            let Some(blueprint) = blueprint else {
+                warn!(&opctx.log,
+                      "Plan execution: skipped";
+                      "reason" => "no blueprint");
+                return json!({"error": "no blueprint" });
+            };
+
+            let result = self.realize_blueprint(opctx, &blueprint).await;
+
+            // Return the result as a `serde_json::Value`
+            match result {
+                Ok(()) => json!({}),
+                Err(errors) => {
+                    let errors: Vec<_> = errors
+                        .into_iter()
+                        .map(|e| format!("{:#}", e))
+                        .collect();
+                    json!({"errors": errors})
+                }
+            }
+        }
+        .boxed()
+    }
+}
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::app::background::common::BackgroundTask;
+    use httptest::matchers::{all_of, json_decoded, request};
+    use httptest::responders::status_code;
+    use httptest::Expectation;
+    use nexus_db_model::{
+        ByteCount, SledBaseboard, SledSystemHardware, SledUpdate,
+    };
+    use nexus_test_utils_macros::nexus_test;
+    use nexus_types::inventory::{
+        OmicronZoneConfig, OmicronZoneDataset, OmicronZoneType,
+    };
+    use omicron_common::api::external::Generation;
+    use serde::Deserialize;
+    use std::collections::BTreeSet;
+    use std::net::SocketAddr;
+
+    type ControlPlaneTestContext =
+        nexus_test_utils::ControlPlaneTestContext<crate::Server>;
+
+    fn create_blueprint(
+        omicron_zones: BTreeMap<Uuid, OmicronZonesConfig>,
+    ) -> Blueprint {
+        Blueprint {
+            id: Uuid::new_v4(),
+            omicron_zones,
+            zones_in_service: BTreeSet::new(),
+            parent_blueprint_id: None,
+            time_created: chrono::Utc::now(),
+            creator: "test".to_string(),
+            comment: "test blueprint".to_string(),
+        }
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_deploy_omicron_zones(cptestctx: &ControlPlaneTestContext) {
```

Comment:
Nice test!

I think we could also/instead have one that uses an actual simulated sled agent and fetches the inventory back to verify it.  I'm not sure there'd be much advantage here, except that eventually we'll probably want to make sure that works so we can test higher-level stuff.  (I wouldn't really worry about this now, just mentioning it.)

Reply:
I particularly didn't do that, because I just wanted to test the behavior of the background task. My hope is that I can get the whole thing running in the a4x2 job and we can use the real sled-agent in CI.